src/cargo/util/toml/mod.rs

```diff
@@ -183,10 +184,33 @@ fn read_manifest_from_str(
             }
         }
     }
+
+    let mut public_deps_without_z = BTreeSet::new();
+    if let Some(deps) = &mut manifest.dependencies {
+        for (name, dep) in deps {
+            if let InheritableDependency::Value(toml_dep) = dep {
+                if toml_dep.is_public() && !config.cli_unstable().public_dependency {
+                    if let TomlDependency::Detailed(d) = toml_dep {
+                        d.public = Some(false)
```

Comment:
Should we set it to `Some(false)` or should we unconditionally set `d.public = None;`?

Reply:
Note that tracking `.is_some()` and switching to `None` would also fix the problem ehuss pointed out

> Is it intentional that public=false does not generate a warning?

Reply:
> I had migrated the code to here, because detailed_dep_to_dependency will invoke by read_manifest_from_str.

These code had deleted.

Reply:
We are now never setting `public = None` in the manifest.  This means that `public` will leak out on `cargo package` / `cargo publish`.

Reply:
Your are right, should set `public = None` here